File: HashMap/Insert Delete GetRandom O(1) - Duplicated Allowed/RandomizedCollection.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Random;

class RandomizedCollection{

    HashMap<Integer, HashSet<Integer>> map; //use hashset to store the multiple indices of a value present in the list
    List<Integer> list;
    Random r;
    
    public RandomizedCollection(){
        map = new HashMap<>();
        list = new ArrayList<>();
        r = new Random();
    }

    public boolean insert(int val){
        boolean flag;
        HashSet<Integer> set;

        //if the value already exists int the map, get the set of indices
        if(map.containsKey(val)){
            set = map.get(val);
            flag = false;
        } else{
            set = new HashSet<>(); //if not, add a new set of indices
            flag = true;
        }

        //add the current size of list as the index for that value
        set.add(list.size());
        //add that value in the list
        list.add(val); 
        //update the map with the new set of indices
        map.put(val, set);

        return flag;
    }

    public boolean remove(int val){
        HashSet<Integer> set;

        //if the value is present in the map, get the set of indices or return false
        if(map.containsKey(val)){
            set = map.get(val);
        } else{
            return false;
        }

        int remIdx = -1;
        for(int i : set){
            remIdx = i; //get the index of one occurence to remove
            break;
        }

        set.remove(remIdx); //remove the index from the set
        map.put(val, set); //update the map with modified set

        //id no occurence of the vale exist -> remove that value from the set
        if(map.get(val).size() == 0){
            map.remove(val);
        }

        //remove the value from the list

        //if the value lies at the last index, remove directly
        if(remIdx == list.size()-1){
            list.remove(list.size()-1);
        } else{
            //if not, swap that index with last element and remove the last element
            list.set(remIdx, list.get(list.size()-1));
            list.remove(list.size()-1);

            //get the value that was moved from the last index in the list
            int newVal = list.get(remIdx);

            //get the set of indices from the map for that value
            HashSet<Integer> temp = map.get(newVal);

            temp.remove(list.size()); //remove the old index of the moved value
            temp.add(remIdx); //add the new index
            map.put(newVal, temp); //update the set int he map
        }

        return true;
    }

    public int getRandom(){ 
        //generate a random number within the range of the list size
        int randomIdx = r.nextInt(list.size());

        //return the value corresponding to that index
        return list.get(randomIdx);
    }
};

File: HashMap/Insert Delete GetRandom O(1) - Duplicated Allowed/RandomizedCollection.cpp
class RandomizedCollection{
public:

    //map where key is the value and the index is the set of indices where that value appears in the list
    unordered_map<int, unordered_set<int>> map;
    vector<int> list;

    RandomizedCollection(){}

    bool insert(int val){
        bool flag;
        unordered_set<int> set; //temporary set to store the indicies where val is found in the list

        //id the value is found in the map
        if(map.find(val) != map.end()){
            set = map[val]; //get the set od indicies where it appears
            flag = false; //it is not the first time inserting this val
        } else{
            flag = true; //it is the first time inserting this val
        }

        //add the index of the new value to the set of that value
        set.insert(list.size());
        list.push_back(val); //add that value to the list
        map[val] = set; //update the map with the modified set for that value

        return flag;
    }

    bool remove(int val){

        //if the value do not present in the map, return false
        if(map.find(val) == map.end()){
            return false;
        }

        //reference the set of indices where the val appears
        auto& set = map[val];
        //get the index of the one occurence of the value to remove it
        int remIdx = *set.begin();

        //remove that index from the set
        set.erase(remIdx);
        //update that set corresponding to the removed value in the map
        map[val] = set;

        //id there are no more occurence of that val, remove it from the map
        if(set.empty()){
            map.erase(val);
        }

        //if the val is present at the last of the list, simply pop it off
        if(remIdx == list.size()-1){
            list.pop_back();
        } else{
            //or, replace that val with the last element and then remove the last element
            list[remidx] = list.back();
            list.pop_back();

            //get the new value replaced from the last
            int newVal = list[remIdx];
            //update the map for the new val's index
            map[newVal].erase(list.size());
            map[newVal].insert(remIdx);
        }

        return true;
    }

    int getRandom(){//get a random index in the range of size of list
        int randomIdx = rand() % list.size();

        //return the value corresponding to that index in the list
        return list[randomIdx];
    }
};